File: cppRobotics/extern/filter/HighPassFilter.cpp
#include "HighPassFilter.h"

static HighPassFilter<double> highPassFilter(50);
static double pParams[2];

double HighPassFilterCalc(double curInput) {
  return highPassFilter.Calc(curInput);
}

double* HighPassFilterGetParams() {
  FirstOrderFilterConfig config = highPassFilter.GetParams();
  pParams[0] = config.cutOffFreqHz;
  pParams[1] = config.timeConstantSec;
  return pParams;
}

void HighPassFilterSetParams(double cutOffFreqHz, double timeConstantSec) {
  FirstOrderFilterConfig config = {cutOffFreqHz, timeConstantSec};
  highPassFilter.SetParams(config);
}

void HighPassFilterResetFilter() {
  highPassFilter.ResetFilter();
}

void HighPassFilterResetPrevValues() {
  highPassFilter.ResetPrevValues();
}

File: lib/filter/high_pass_filter.dart
import '../native_cpp.dart';
import 'dart:ffi'; // For FFI

/// T curOutput = alpha_ * prevOutput_ + (1.0 - alpha_) * (curInput - prevInput_);
final double Function(double curInput) highPassFilterCalc = cppRoboticsLib
    .lookup<NativeFunction<Double Function(Double)>>('HighPassFilterCalc')
    .asFunction();

/// params[0] = config.cutOffFreqHz;
///
/// params[1] = config.timeConstantSec;
final Pointer<Double> Function() highPassFilterGetParams = cppRoboticsLib
    .lookup<NativeFunction<Pointer<Double> Function()>>(
        'HighPassFilterGetParams')
    .asFunction();

final void Function(double cutOffFreqHz, double timeConstantSec)
    highPassFilterSetParams = cppRoboticsLib
        .lookup<NativeFunction<Void Function(Double, Double)>>(
            'HighPassFilterSetParams')
        .asFunction();

final void Function() highPassFilterResetFilter = cppRoboticsLib
    .lookup<NativeFunction<Void Function()>>('HighPassFilterResetFilter')
    .asFunction();

final void Function() highPassFilterResetPrevValues = cppRoboticsLib
    .lookup<NativeFunction<Void Function()>>('HighPassFilterResetPrevValues')
    .asFunction();
